File: examples/bot.py
"""
War of Agents — Python Agent Bot Example
Connects to the game server, registers a hero, and plays autonomously.

Usage:
    pip install requests websocket-client
    python bot.py
"""

import requests
import json
import time
import random
import websocket
import threading

SERVER = "http://localhost:3001"
WS_SERVER = "ws://localhost:3001"

# ─── Configuration ───────────────────────────────────────────
AGENT_ID = f"py-bot-{random.randint(1000,9999)}"
AGENT_NAME = "PythonBot"
FACTION = random.choice(["alliance", "horde"])
HERO_CLASS = random.choice(["knight", "ranger", "mage", "priest", "siegemaster"])

game_state = None
hero_id = None


def register():
    """Register agent with the server."""
    global hero_id
    res = requests.post(f"{SERVER}/api/agents/register", json={
        "agentId": AGENT_ID,
        "name": AGENT_NAME,
        "faction": FACTION,
        "heroClass": HERO_CLASS,
    })
    data = res.json()
    if data.get("success"):
        hero_id = data["heroId"]
        print(f"Registered: {AGENT_NAME} ({FACTION} {HERO_CLASS}) -> {hero_id}")
    else:
        print(f"Registration failed: {data}")
        exit(1)


def get_my_hero():
    """Find our hero in the game state."""
    if not game_state:
        return None
    for h in game_state.get("heroes", []):
        if h.get("agentId") == AGENT_ID:
            return h
    return None


def deploy(action, **kwargs):
    """Send a strategy deployment command."""
    payload = {"agentId": AGENT_ID, "action": action, **kwargs}
    try:
        res = requests.post(f"{SERVER}/api/strategy/deployment", json=payload)
        return res.json()
    except Exception as e:
        print(f"Deploy error: {e}")
        return {}


def find_nearest_enemy(hero):
    """Find the closest enemy unit or hero."""
    hx, hy = hero["x"], hero["y"]
    best = None
    best_dist = float("inf")

    for u in game_state.get("units", []):
        if u["faction"] != FACTION and u.get("alive", True):
            dx, dy = u["x"] - hx, u["y"] - hy
            d = (dx*dx + dy*dy) ** 0.5
            if d < best_dist:
                best_dist = d
                best = u

    for h in game_state.get("heroes", []):
        if h["faction"] != FACTION and h.get("alive", True):
            dx, dy = h["x"] - hx, h["y"] - hy
            d = (dx*dx + dy*dy) ** 0.5
            if d < best_dist:
                best_dist = d
                best = h

    return best, best_dist


def think():
    """Main AI loop — called every second."""
    hero = get_my_hero()
    if not hero or not hero.get("alive"):
        return

    # Retreat if low HP
    if hero["hp"] < hero["maxHp"] * 0.25:
        base_x = 150 if FACTION == "alliance" else 4650
        deploy("move", targetX=base_x, targetY=1200)
        print(f"[{AGENT_NAME}] Retreating! HP: {hero['hp']}/{hero['maxHp']}")
        return

    # Find nearest enemy
    enemy, dist = find_nearest_enemy(hero)

    if enemy:
        # Move toward enemy
        deploy("move", targetX=enemy["x"], targetY=enemy["y"])

        # Use abilities if in range
        for ab in hero.get("abilities", []):
            if ab["cd"] == 0:
                deploy("ability", abilityId=ab["id"])
                print(f"[{AGENT_NAME}] Cast {ab['name']}!")
                break

    # Buy items if we have gold
    if hero["gold"] >= 300:
        for item_id in ["boots", "sword", "shield", "cloak", "relic"]:
            res = deploy("buy", itemId=item_id)
            if res.get("success"):
                print(f"[{AGENT_NAME}] Bought {res.get('item', item_id)}!")
                break


def on_ws_message(ws, message):
    """Handle WebSocket state updates."""
    global game_state
    msg = json.loads(message)
    if msg.get("type") == "state":
        game_state = msg["data"]


def on_ws_error(ws, error):
    print(f"WS Error: {error}")


def on_ws_close(ws, code, msg):
    print("WS Disconnected")


def on_ws_open(ws):
    print(f"Connected to {WS_SERVER}")


def main():
    print(f"=== War of Agents — Python Bot ===")
    print(f"Server: {SERVER}")

    # Register
    register()

    # Connect WebSocket for state updates
    ws = websocket.WebSocketApp(
        WS_SERVER,
        on_message=on_ws_message,
        on_error=on_ws_error,
        on_close=on_ws_close,
        on_open=on_ws_open,
    )
    ws_thread = threading.Thread(target=ws.run_forever, daemon=True)
    ws_thread.start()

    time.sleep(1)

    # Main loop
    print(f"Bot running as {FACTION} {HERO_CLASS}...")
    while True:
        try:
            think()
        except Exception as e:
            print(f"Error: {e}")
        time.sleep(1)


if __name__ == "__main__":
    main()

File: examples/bot.js
/**
 * War of Agents — JavaScript Agent Bot Example
 * Connects to the game server, registers a hero, and plays autonomously.
 *
 * Usage:
 *   npm install ws
 *   node bot.js
 */

const WebSocket = require('ws');
const http = require('http');
const https = require('https');

const SERVER = process.env.SERVER || 'http://localhost:3001';
const WS_SERVER = SERVER.replace('http', 'ws');

const AGENT_ID = `js-bot-${Math.floor(Math.random() * 9000) + 1000}`;
const AGENT_NAME = 'JSBot';
const FACTION = Math.random() > 0.5 ? 'alliance' : 'horde';
const HERO_CLASS = ['knight', 'ranger', 'mage', 'priest', 'siegemaster'][Math.floor(Math.random() * 5)];

let gameState = null;

// ─── HTTP helpers ────────────────────────────────────────────
function post(path, body) {
  return new Promise((resolve, reject) => {
    const url = new URL(path, SERVER);
    const mod = url.protocol === 'https:' ? https : http;
    const data = JSON.stringify(body);
    const req = mod.request(url, {
      method: 'POST',
      headers: { 'Content-Type': 'application/json', 'Content-Length': data.length },
    }, (res) => {
      let buf = '';
      res.on('data', c => buf += c);
      res.on('end', () => { try { resolve(JSON.parse(buf)); } catch { resolve(buf); } });
    });
    req.on('error', reject);
    req.write(data);
    req.end();
  });
}

// ─── Register ────────────────────────────────────────────────
async function register() {
  const res = await post('/api/agents/register', {
    agentId: AGENT_ID,
    name: AGENT_NAME,
    faction: FACTION,
    heroClass: HERO_CLASS,
  });
  if (res.success) {
    console.log(`Registered: ${AGENT_NAME} (${FACTION} ${HERO_CLASS}) -> ${res.heroId}`);
  } else {
    console.error('Registration failed:', res);
    process.exit(1);
  }
}

// ─── Find our hero ───────────────────────────────────────────
function getMyHero() {
  if (!gameState) return null;
  return gameState.heroes.find(h => h.agentId === AGENT_ID) || null;
}

// ─── Deploy command ──────────────────────────────────────────
async function deploy(action, extra = {}) {
  return post('/api/strategy/deployment', { agentId: AGENT_ID, action, ...extra });
}

// ─── Find nearest enemy ─────────────────────────────────────
function findNearestEnemy(hero) {
  let best = null, bestDist = Infinity;
  const check = (list) => {
    for (const e of list) {
      if (e.faction === FACTION || !e.alive) continue;
      const dx = e.x - hero.x, dy = e.y - hero.y;
      const d = Math.sqrt(dx * dx + dy * dy);
      if (d < bestDist) { bestDist = d; best = e; }
    }
  };
  check(gameState.units || []);
  check(gameState.heroes || []);
  return { enemy: best, dist: bestDist };
}

// ─── AI Think Loop ───────────────────────────────────────────
async function think() {
  const hero = getMyHero();
  if (!hero || !hero.alive) return;

  // Retreat if low HP
  if (hero.hp < hero.maxHp * 0.25) {
    const baseX = FACTION === 'alliance' ? 150 : 4650;
    await deploy('move', { targetX: baseX, targetY: 1200 });
    console.log(`[${AGENT_NAME}] Retreating! HP: ${hero.hp}/${hero.maxHp}`);
    return;
  }

  // Find nearest enemy
  const { enemy, dist } = findNearestEnemy(hero);

  if (enemy) {
    await deploy('move', { targetX: enemy.x, targetY: enemy.y });

    // Use abilities
    for (const ab of (hero.abilities || [])) {
      if (ab.cd === 0) {
        await deploy('ability', { abilityId: ab.id });
        console.log(`[${AGENT_NAME}] Cast ${ab.name}!`);
        break;
      }
    }
  }

  // Buy items
  if (hero.gold >= 300) {
    for (const itemId of ['boots', 'sword', 'shield', 'cloak', 'relic']) {
      const res = await deploy('buy', { itemId });
      if (res.success) {
        console.log(`[${AGENT_NAME}] Bought ${res.item || itemId}!`);
        break;
      }
    }
  }
}

// ─── Main ────────────────────────────────────────────────────
async function main() {
  console.log('=== War of Agents — JS Bot ===');
  console.log(`Server: ${SERVER}`);

  await register();

  // WebSocket for state updates
  const ws = new WebSocket(WS_SERVER);
  ws.on('message', (data) => {
    const msg = JSON.parse(data.toString());
    if (msg.type === 'state') gameState = msg.data;
  });
  ws.on('open', () => console.log('WebSocket connected'));
  ws.on('error', (e) => console.error('WS error:', e.message));

  // Wait for first state
  await new Promise(r => setTimeout(r, 1500));

  console.log(`Bot running as ${FACTION} ${HERO_CLASS}...`);

  // Think loop every second
  setInterval(async () => {
    try { await think(); } catch (e) { console.error('Error:', e.message); }
  }, 1000);
}

main();
